src/main/java/com/diffplug/gradle/spotless/FormatTask.java

```diff
@@ -23,19 +23,36 @@
 import java.util.List;
 import java.util.Locale;
 
+import com.diffplug.common.base.Errors;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFiles;
+import org.gradle.api.tasks.SkipWhenEmpty;
 import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.incremental.IncrementalTaskInputs;
 
 public class FormatTask extends DefaultTask {
 	// set by SpotlessExtension, but possibly overridden by FormatExtension
+	@Input
 	public Charset encoding = StandardCharsets.UTF_8;
-	public LineEnding.Policy lineEndingsPolicy = LineEnding.UNIX_POLICY;
+	@Input
+	public LineEnding.Policy lineEndingPolicy = LineEnding.UNIX_POLICY;
+
 	// set by FormatExtension
+	@Input
 	public boolean paddedCell = false;
+	@InputFiles
+	@OutputFiles
+	@SkipWhenEmpty
 	public Iterable<File> target;
+	@Input
```

Comment:
> You can write a custom rule by implementing such-and-such class and then calling the constructor here, and if you change the rule be sure to change the serialVersionUID


I think you misunderstood my point above.

- if the author implements an inline rule it will *always* be out-of-date and thus *correct*
- if the author wants extra performance, they can write a class with equals/hashcode

This gives you the best of both worlds: Ease of use by default, performance on demand. No need for a 2-task solution.